src/App.js

```diff
@@ -1,86 +1,201 @@
-import React from 'react';
+import React, { useState, useEffect } from 'react';
+import { v4 as uuidv4 } from 'uuid';
+import { TodoList } from './components/TodoList';
+import { Filters } from './components/Filters';
+
+const useLocalStorage = (key, initialValue) => {
+  const [value, setValue] = useState(
+    JSON.parse(localStorage.getItem(key)) || initialValue,
+  );
+  const save = (newValue) => {
+    setValue(newValue);
+    localStorage.setItem(key, JSON.stringify(newValue));
+  };
+
+  return [value, save];
+};
 
 function App() {
+  const [title, setTitle] = useState('');
+  const [todos, setTodos] = useLocalStorage('todos', []);
+  const [filteredTodos, setfilteredTodos] = useState(todos);
+  const [filterStatus, setFilterStatus] = useState('');
+  const [activeSelectAll, setActiveSelectAll] = useState(false);
+
+  const addTodo = (value) => {
+    setTitle(value);
+  };
+
+  const onSubmit = (event) => {
+    event.preventDefault();
+
+    if (title.length === 0) {
+      return;
+    }
+
+    const newTodo = {
+      title,
+      id: uuidv4(),
+      completed: false,
+    };
+
+    setTodos([...todos, newTodo]);
+    setfilteredTodos([...todos, newTodo]);
+
+    setTitle('');
+  };
+
+  const toggleAll = () => {
+    const uncheckedTodos = todos.map(todo => (
+      {
+        ...todo,
+        completed: false,
+      }
+    ));
+
+    setTodos(uncheckedTodos);
+    setfilteredTodos(uncheckedTodos);
+  };
+
+  const changeStatus = (checkedTodoId) => {
+    const todosCopy = todos.map((todo) => {
+      if (todo.id === checkedTodoId) {
+        return {
+          ...todo,
+          completed: !todo.completed,
+        };
+      }
+
+      return todo;
+    });
+
+    setTodos(todosCopy);
+    setfilteredTodos(todosCopy);
+  };
+
+  useEffect(() => {
+    if (filterStatus === 'all' || filterStatus === '') {
+      setfilteredTodos(todos);
+    } else {
+      const filter = filterStatus === 'completed';
+
+      const filteredList = todos.filter(todo => (
+        todo.completed === filter
+      ));
+
+      setfilteredTodos(filteredList);
+    }
+  }, [filterStatus, todos]);
+
+  const deleteTodo = (todoIdForDelete) => {
+    const filteredList = todos.filter(todo => (
+      todo.id !== todoIdForDelete
+    ));
+
+    setTodos(filteredList);
+    setfilteredTodos(filteredList);
+  };
+
+  useEffect(() => {
+    if (!activeSelectAll) {
+      const checkStatus = todos.some(todo => (
+        todo.completed === true
+      ));
+
+      if (checkStatus) {
+        setActiveSelectAll(true);
+      }
+    } else {
+      const checkStatus = todos.every(todo => (
+        todo.completed === false
+      ));
+
+      if (checkStatus) {
+        setActiveSelectAll(false);
+      }
+    }
+  }, [todos, activeSelectAll]);
```

Comment:
what is happening here? 0_0

Reply:
The Footer component was updated for the same action (but with a shorter code).